src/hotspot/share/opto/parse1.cpp

```diff
@@ -921,7 +921,7 @@ void Compile::return_values(JVMState* jvms) {
     kit.inc_sp(-ret_size);  // pop the return value(s)
     kit.sync_jvms();
     Node* res = kit.argument(0);
-    if (res->isa_InlineType() && VectorSupport::skip_value_scalarization(res->as_InlineType()->inline_klass())) {
+    if (res->isa_InlineType() && VectorSupport::skip_value_scalarization(res->as_InlineType()->inline_klass()->get_InlineKlass())) {
```

Comment:
I'm wondering if `tf()->returns_inline_type_as_fields()` in followed `else-if` branch has any connection with `ciInlineKlass::can_be_returned_as_fields()`.  If `ciInlineKlass::can_be_returned_as_fields()` can decide the result of `tf()->returns_inline_type_as_fields()`, then this special handling for bufferred `InlineTypeNode` can be removed like before?

Reply:
C2 always scalarizes value objects before returning it, for vector we current do not support scalarization hence I am returning the oop buffer of InlineTypeNode.

Reply:
BTW,  Valhalla  added a new return calling convention for inline types but it needs to be extended to cover vectors.
[https://github.com/openjdk/valhalla/blob/lworld/src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp#L567](https://github.com/openjdk/valhalla/blob/lworld/src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp#L567)

Reply:
Yeah, make sense to me. What I mean here is whether we can save this new added branch if  the `else-if` in line-928 can avoid such multifields scalarization. Not a block, we can refine such code in future if possible.